pallets/disputes/src/lib.rs

```diff
@@ -0,0 +1,382 @@
+        type MaxReasonLength: Get<u32>;
+        /// This is number of juries that can be assigned to a given dispute
+        type MaxJurySize: Get<u32>;
+        /// This is number of specifics that can be assigned to a given dispute
+        type MaxSpecifics: Get<u32>;
+        /// The amount of time a dispute takes to finalise.
+        type VotingTimeLimit: Get<<Self as frame_system::Config>::BlockNumber>;
+        /// The origin used to force cancel and pass disputes.
+        type ForceOrigin: EnsureOrigin<Self::RuntimeOrigin>;
+        /// External hooks to handle the completion of a dispute.
+        type DisputeHooks: DisputeHooks<Self::DisputeKey>;
+    }
+
+    /// Used to store the disputes that is being raised, given the dispute key it returns the Dispute
+    /// Key: DisputeKey
+    /// Value: Dispute<T>
+    #[pallet::storage]
+    #[pallet::getter(fn disputes)]
+    pub type Disputes<T: Config> =
+        StorageMap<_, Blake2_128Concat, T::DisputeKey, Dispute<T>, OptionQuery>;
+
+    /// Stores the dispute keys that will finalise on a given block.
+    /// Key: BlockNumber
+    /// Value: Vec<DisputeKey>
+    #[pallet::storage]
+    pub type DisputesFinaliseOn<T: Config> = StorageMap<
+        _,
+        Blake2_128Concat,
+        BlockNumberFor<T>,
+        BoundedVec<T::DisputeKey, ConstU32<1000>>,
+        ValueQuery,
+    >;
+
+    #[pallet::event]
+    #[pallet::generate_deposit(pub(super) fn deposit_event)]
+    pub enum Event<T: Config> {
+        /// A dispute has been raised.
+        DisputeRaised {
+            dispute_key: T::DisputeKey,
+        },
+        /// A disute has been voted on
+        DisputeVotedOn {
+            who: AccountIdOf<T>,
+        },
+        /// A dispute has been completed.
+        // TODO: Not in use
+        DisputeCompleted {
+            dispute_key: T::DisputeKey,
+        },
+        /// A dispute has been cancelled.
+        // TODO: Not in use
+        DisputeCancelled {
+            dispute_key: T::DisputeKey,
+        },
+        /// A dispute has been extended.
+        DisputeExtended {
+            dispute_key: T::DisputeKey,
+        },
+    }
+
+    #[pallet::error]
+    pub enum Error<T> {
+        /// Dispute key does not exist.
+        DisputeDoesNotExist,
+        /// Dispute key already exists.
+        DisputeAlreadyExists,
+        // This account is not part of the specified jury.
+        NotAJuryAccount,
+        /// There have been too many disputes on this block. Try next block.
+        TooManyDisputesThisBlock,
+        /// The dispute has already been extended. You can only extend a dispute once.
+        DisputeAlreadyExtended,
+        /// There have been more than required votes for a given dispute
+        TooManyDisputeVotes,
+    }
+
+    #[pallet::hooks]
+    impl<T: Config> Hooks<BlockNumberFor<T>> for Pallet<T> {
+        // TODO: WEIGHT + BENCHMARKS
+        fn on_initialize(n: BlockNumberFor<T>) -> Weight {
```

Comment:
None of this has been used in the test. 
What what you should do to test these is: 
1. Start a dispute
2. vote without triggering unanimous autofinalisation
3. wait till the finalisation block
4. check to see if it passes or fails (test for both cases)